src/backend/distributed/operations/shard_rebalancer.c

```diff
@@ -429,6 +432,60 @@ GetShardCost(uint64 shardId, void *voidContext)
 }
 
 
+/*
+ * citus_partitioned_shard_total_size returns the total size of a partitioned table with
+ * its partitions, using the function pg_partition_tree.
+ */
+Datum
+citus_partitioned_shard_total_size(PG_FUNCTION_ARGS)
+{
+	char *shardName = text_to_cstring(PG_GETARG_TEXT_P(0));
+	uint64 size = 0;
+	const int paramCount = 1;
+	Oid paramTypes[1] = { TEXTOID };
+	Datum paramValues[1] = { CStringGetTextDatum(shardName) };
+
+	/* pg_partition_tree finds the partitions recursively and returns them */
+	char *queryText = "SELECT sum(pg_total_relation_size(relid))::bigint AS total_size "
+					  "FROM (SELECT relid from pg_partition_tree($1)) tables;";
+
+	int spiResult = SPI_connect();
+	if (spiResult != SPI_OK_CONNECT)
+	{
+		ereport(ERROR, (errmsg("could not connect to SPI manager")));
+	}
+
+	spiResult = SPI_execute_with_args(queryText, paramCount, paramTypes,
+									  paramValues,
+									  NULL, false, 0);
+
+	if (spiResult == SPI_OK_SELECT)
+	{
+		Assert(SPI_processed <= 1);
+		for (uint64 row = 0; row < SPI_processed; row++)
+		{
+			bool isnull = false;
+
+			size = DatumGetInt64(SPI_getbinval(SPI_tuptable->vals[0],
+											   SPI_tuptable->tupdesc,
+											   1, &isnull));
+		}
+	}
+	else
+	{
+		ereport(ERROR, (errmsg("could not run SPI query")));
```

Comment:
same as above These errors boil back to the user in most cases, using things like SPI will tell them nothing.